### PR TITLE
Delete Workspace

### PR DESCRIPTION
**Pull request by yuye-aws**
### Description

Implement delete on workspace overview page. Users can now delete the current workspace.

## Screenshot

image

### Check List

- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [x] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by Hailong-am** [comment]
the PR target branch should be [ruanyl:workspace](https://github.com/ruanyl/OpenSearch-Dashboards/tree/workspace)

**Comment by yuye-aws** [comment]
> the PR target branch should be [ruanyl:workspace](https://github.com/ruanyl/OpenSearch-Dashboards/tree/workspace)

Sorry, I have changed.

File: src/plugins/workspace/public/components/workspace_overview.tsx [hunk 1]
```diff
@@ -3,32 +3,53 @@
...
     workspaces ? workspaces.client.currentWorkspace$ : of(null)
   );
 
+  const workspaceId = currentWorkspace?.id;
+  const workspaceName = currentWorkspace?.name;
+  const [deleteWorkspaceModalVisible, setDeleteWorkspaceModalVisible] = useState(false);
+
+  const deleteWorkspace = () => {
+    if (workspaceId) {
+      workspaces?.client.delete(workspaceId);
```
Comment (SuZhou-Joe) [inline src/plugins/workspace/public/components/workspace_overview.tsx:29]: delete should be a async call and we should show a error toast if this API failed.

File: src/plugins/workspace/public/components/workspace_overview.tsx [hunk 2]
```diff
@@ -3,32 +3,53 @@
...
-          <EuiButton color="danger">Delete</EuiButton>,
+          <EuiButton color="danger" onClick={() => setDeleteWorkspaceModalVisible(true)}>
+            Delete
+          </EuiButton>,
           <EuiButton>Update</EuiButton>,
         ]}
       />
       <EuiPanel>
+        <DeleteWorkspaceModal
+          onConfirm={() => deleteWorkspace()}
```
Comment (SuZhou-Joe) [inline src/plugins/workspace/public/components/workspace_overview.tsx:48]:
```suggestion
          onConfirm={deleteWorkspace}
```

File: src/plugins/workspace/public/components/workspace_overview.tsx [hunk 3]
```diff
@@ -3,32 +3,53 @@
...
+          onClose={() => setDeleteWorkspaceModalVisible(false)}
+          visible={deleteWorkspaceModalVisible}
+          selectedItems={[workspaceName ?? 'null']}
```
Comment (SuZhou-Joe) [inline src/plugins/workspace/public/components/workspace_overview.tsx:51]: What does a string "null" mean here, I do not see any consuming logic in DeleteWorkspaceModal. Maybe a empty [] is enough?

File: src/plugins/workspace/public/components/delete_workspace_modal/delete_workspace_modal.tsx [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+  selectedItems: string[];
+  visible: boolean;
+  onClose: () => void;
+  onConfirm: () => void;
+}
+
+// eslint-disable-next-line import/no-default-export
+export default function DeleteWorkspaceModal(props: DeleteWorkspaceModalProps) {
+  const [value, setValue] = useState('');
+  const { onClose, onConfirm, visible, selectedItems } = props;
```
Comment (ruanyl) [inline src/plugins/workspace/public/components/delete_workspace_modal/delete_workspace_modal.tsx:30]: I think it's better to control the visibility of DeleteWorkspaceModal at `workspace_overview.tsx`, it should be like 
```typescript
{deleteWorkspaceModalVisible && <DeleteWorkspaceModal />}
```

**Comment by yuye-aws** [comment]
Thanks for the comments! I have pushed three new commits. Please help review.

File: src/plugins/workspace/public/components/delete_workspace_modal/delete_workspace_modal.tsx [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+interface DeleteWorkspaceModalProps {
+  selectedItems: string[];
+  onClose: () => void;
+  onConfirm: () => void;
+}
+
+export function DeleteWorkspaceModal(props: DeleteWorkspaceModalProps) {
+  const [value, setValue] = useState('');
+  const { onClose, onConfirm, selectedItems } = props;
+
```
Comment (SuZhou-Joe) [inline src/plugins/workspace/public/components/delete_workspace_modal/delete_workspace_modal.tsx:29]: There was a useEffect here to reset the input value, it should not have been deleted.

File: src/plugins/workspace/public/components/workspace_overview.tsx [hunk 6]
```diff
@@ -3,32 +3,80 @@
...
+        notifications?.toasts.addDanger({
+          title: i18n.translate('workspace.delete.failed', {
+            defaultMessage: 'Failed to delete workspace',
+          }),
+          text: result?.error,
+        });
+      }
+    }
+    setDeleteWorkspaceModalVisible(false);
+    await application.navigateToApp('home');
```
Comment (SuZhou-Joe) [inline src/plugins/workspace/public/components/workspace_overview.tsx:58]: I am not sure if there is a enum for core App name, maybe we can take a look in core dashboard repo?

Comment (yuye-aws) [reply src/plugins/workspace/public/components/workspace_overview.tsx:58]: Currently there is no enum for 'home', i think this part of the code can remain unchanged.

**Review by SuZhou-Joe — APPROVED** [review]

File: src/plugins/workspace/public/components/delete_workspace_modal/delete_workspace_modal.tsx (hunk 5 above)
Comment (yuye-aws) [reply src/plugins/workspace/public/components/delete_workspace_modal/delete_workspace_modal.tsx:29]: In workspace overview page, I add a variable named "deleteWorkspaceModalVisible" to control the visibility of the delete modal:  
`{deleteWorkspaceModalVisible && <DeleteWorkspaceModal />}`
Therefore, the input value will be reset when the delete modal changes visibility.

**Review by ruanyl — APPROVED** [review]

File: src/plugins/workspace/public/components/delete_workspace_modal/delete_workspace_modal.tsx [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+  return (
+    <EuiModal onClose={onClose}>
```
Comment (ruanyl) [inline src/plugins/workspace/public/components/delete_workspace_modal/delete_workspace_modal.tsx:31]: We have `EuiConfirmModal`, you can actually use that

**Pull request merged**